File: src/llaisys/models/qwen2.cc
#include "llaisys/models/qwen2.h"

#include "../llaisys_tensor.hpp"

#include "../../tensor/tensor.hpp"

#include "../../utils.hpp"

#include <random>
#include <vector>
#include <algorithm>
#include <cmath>

#include "../../ops/add/op.hpp"
#include "../../ops/argmax/op.hpp"
#include "../../ops/embedding/op.hpp"
#include "../../ops/linear/op.hpp"
#include "../../ops/rearrange/op.hpp"
#include "../../ops/rms_norm/op.hpp"
#include "../../ops/rope/op.hpp"
#include "../../ops/self_attention/op.hpp"
#include "../../ops/swiglu/op.hpp"

//实现Qwen2的api

__C{
    struct LlaisysQwen2Model{
        LlaisysQwen2Meta meta;
        LlaisysQwen2Weights* weights;

        llaisysDeviceType_t device;
        int *device_ids;
        int ndevice;

        //定义kv cache相关
        llaisysTensor_t *k_cache;
        llaisysTensor_t *v_cache;

        // size_t cached = 0;
    };

    __export struct LlaisysQwen2Model *llaisysQwen2ModelCreate(const LlaisysQwen2Meta *meta, const size_t kv_cache_size,
        llaisysDeviceType_t device, int *device_ids, int ndevice){
        //初始化模型
        LOG("llaisysQwen2ModelCreate");
        LlaisysQwen2Model *model = new LlaisysQwen2Model();
        model->meta = *meta;
        model->device = device;
        //深拷贝device_ids
        model->device_ids = new int[ndevice];
        std::copy(device_ids, device_ids + ndevice, model->device_ids);
        model->ndevice = ndevice;
        
        //分配weights结构体
        model->weights = new LlaisysQwen2Weights();

        //初始化kv cache
        size_t nlayer = meta->nlayer;
        model->k_cache = new llaisysTensor_t[nlayer];
        model->v_cache = new llaisysTensor_t[nlayer];
        size_t kv_shape[3] = {kv_cache_size, meta->nkvh, meta->dh};
        for(size_t i = 0; i < nlayer; i++){
            model->k_cache[i] = tensorCreate(kv_shape, 3, meta->dtype, device, device_ids[0]);
            model->v_cache[i] = tensorCreate(kv_shape, 3, meta->dtype, device, device_ids[0]);
        }
        LOG("llaisysQwen2ModelCreate: k_cache");
        LOG("llaisysQwen2ModelCreate: v_cache");


        //初始化其他权重
        LOG("llaisysQwen2ModelCreate: in_embed");
        std::vector<size_t> in_embed_shape = {meta->voc, meta->hs};
        model->weights->in_embed = tensorCreate(in_embed_shape.data(), 2, meta->dtype, device, device_ids[0]);
        LOG("llaisysQwen2ModelCreate: out_embed");
        std::vector<size_t> out_embed_shape = {meta->voc, meta->hs};
        model->weights->out_embed = tensorCreate(out_embed_shape.data(), 2, meta->dtype, device, device_ids[0]);
        LOG("llaisysQwen2ModelCreate: out_norm_w");
        std::vector<size_t> out_norm_w_shape = {meta->hs};
        model->weights->out_norm_w = tensorCreate(out_norm_w_shape.data(), 1, meta->dtype, device, device_ids[0]);

        //初始化其他权重
        model->weights->attn_norm_w = new llaisysTensor_t[nlayer];
        model->weights->attn_q_w = new llaisysTensor_t[nlayer];
        model->weights->attn_k_w = new llaisysTensor_t[nlayer];
        model->weights->attn_v_w = new llaisysTensor_t[nlayer];
        model->weights->attn_o_w = new llaisysTensor_t[nlayer];
        model->weights->attn_q_b = new llaisysTensor_t[nlayer];
        model->weights->attn_k_b = new llaisysTensor_t[nlayer];
        model->weights->attn_v_b = new llaisysTensor_t[nlayer];
        model->weights->mlp_up_w = new llaisysTensor_t[nlayer];
        model->weights->mlp_down_w = new llaisysTensor_t[nlayer];
        model->weights->mlp_gate_w = new llaisysTensor_t[nlayer];
        model->weights->mlp_norm_w = new llaisysTensor_t[nlayer];

        size_t attn_norm_w_shape[1] = {meta->hs};
        //hs = nh * dh
        size_t attn_q_w_shape[2] = {meta->hs, meta->hs};
        size_t attn_k_w_shape[2] = {meta->nkvh * meta->dh, meta->hs};
        size_t attn_v_w_shape[2] = {meta->nkvh * meta->dh, meta->hs};
        size_t attn_o_w_shape[2] = {meta->hs, meta->hs};

        size_t attn_q_b_shape[1] = {meta->hs};
        size_t attn_k_b_shape[1] = {meta->nkvh * meta->dh};
        size_t attn_v_b_shape[1] = {meta->nkvh * meta->dh};

        size_t mlp_down_w_shape[2] = {meta->hs, meta->di};
        size_t mlp_up_w_shape[2] = {meta->di, meta->hs};
        size_t mlp_gate_w_shape[2] = {meta->di, meta->hs};
        size_t mlp_norm_w_shape[1] = {meta->hs};

        for(size_t i = 0; i < nlayer; i++){
            model->weights->attn_norm_w[i] = tensorCreate(attn_norm_w_shape, 1, meta->dtype, device, device_ids[0]);
            model->weights->attn_q_w[i] = tensorCreate(attn_q_w_shape, 2, meta->dtype, device, device_ids[0]);
            model->weights->attn_k_w[i] = tensorCreate(attn_k_w_shape, 2, meta->dtype, device, device_ids[0]);
            model->weights->attn_v_w[i] = tensorCreate(attn_v_w_shape, 2, meta->dtype, device, device_ids[0]);
            model->weights->attn_o_w[i] = tensorCreate(attn_o_w_shape, 2, meta->dtype, device, device_ids[0]);
            model->weights->attn_q_b[i] = tensorCreate(attn_q_b_shape, 1, meta->dtype, device, device_ids[0]);
            model->weights->attn_k_b[i] = tensorCreate(attn_k_b_shape, 1, meta->dtype, device, device_ids[0]);
            model->weights->attn_v_b[i] = tensorCreate(attn_v_b_shape, 1, meta->dtype, device, device_ids[0]);
            model->weights->mlp_down_w[i] = tensorCreate(mlp_down_w_shape, 2, meta->dtype, device, device_ids[0]);
            model->weights->mlp_up_w[i] = tensorCreate(mlp_up_w_shape, 2, meta->dtype, device, device_ids[0]);
            model->weights->mlp_gate_w[i] = tensorCreate(mlp_gate_w_shape, 2, meta->dtype, device, device_ids[0]);
            model->weights->mlp_norm_w[i] = tensorCreate(mlp_norm_w_shape, 1, meta->dtype, device, device_ids[0]);
        }
        LOG("model created");

        return model;
    }

    __export void llaisysQwen2ModelDestroy(struct LlaisysQwen2Model * model){
        if(model == nullptr){
            return;
        }

        size_t nlayer = model->meta.nlayer;
        //销毁kv cache
        for(size_t i = 0; i < nlayer; i++){
            tensorDestroy(model->k_cache[i]);
            tensorDestroy(model->v_cache[i]);
        }
        delete[] model->k_cache;
        delete[] model->v_cache;
        
        //销毁其他权重
        tensorDestroy(model->weights->in_embed);
        tensorDestroy(model->weights->out_embed);
        tensorDestroy(model->weights->out_norm_w);
        for(size_t i = 0; i < nlayer; i++){
            tensorDestroy(model->weights->attn_norm_w[i]);
            tensorDestroy(model->weights->attn_q_w[i]);
            tensorDestroy(model->weights->attn_k_w[i]);
            tensorDestroy(model->weights->attn_v_w[i]);
            tensorDestroy(model->weights->attn_o_w[i]);
            tensorDestroy(model->weights->attn_q_b[i]);
            tensorDestroy(model->weights->attn_k_b[i]);
            tensorDestroy(model->weights->attn_v_b[i]);
            tensorDestroy(model->weights->mlp_down_w[i]);
            tensorDestroy(model->weights->mlp_up_w[i]);
            tensorDestroy(model->weights->mlp_gate_w[i]);
            tensorDestroy(model->weights->mlp_norm_w[i]);
        }
        //销毁其他权重
        delete[] model->weights->attn_norm_w;
        delete[] model->weights->attn_q_w;
        delete[] model->weights->attn_k_w;
        delete[] model->weights->attn_v_w;
        delete[] model->weights->attn_o_w;
        delete[] model->weights->attn_q_b;
        delete[] model->weights->attn_k_b;
        delete[] model->weights->attn_v_b;
        delete[] model->weights->mlp_down_w;
        delete[] model->weights->mlp_up_w;
        delete[] model->weights->mlp_gate_w;
        delete[] model->weights->mlp_norm_w;
        
        //销毁weights结构体
        delete model->weights;

        if(model -> device_ids != nullptr){
            delete[] model -> device_ids;
        }
        delete model;
    }

    __export struct LlaisysQwen2Weights *llaisysQwen2ModelWeights(struct LlaisysQwen2Model * model){
        // TO_BE_IMPLEMENTED();
        if(model != nullptr){
            return model->weights;
        }
        return nullptr;
    }

    __export int64_t llaisysQwen2ModelInfer(struct LlaisysQwen2Model * model, int64_t * token_ids, size_t ntoken, size_t past_len){
        /*
            reference:https://github.com/zebra-uestc/llaisys/blob/master/src/llaisys/models/qwen2.cc\
        */
        // TO_BE_IMPLEMENTED();
        if(model == nullptr || token_ids == nullptr || ntoken == 0){
            std::cerr << "model or token_ids is nullptr" << std::endl;
            return -1;
        }
        
        //weights
        if(model -> weights->in_embed == nullptr || 
            model -> weights->out_embed == nullptr ||
            model -> weights->out_norm_w == nullptr){
            std::cerr << "in_embed or out_embed or out_norm_w is nullptr" << std::endl;
            return -1;
        }

        //layer weights
        int device_id = 0;
        if(model -> device_ids != nullptr && model -> ndevice > 0){
            device_id = model -> device_ids[0];
        }
        auto device_type = model->device;

        //some variables
        size_t nlayer = model->meta.nlayer;
        size_t voc = model->meta.voc;
        size_t di = model->meta.di;
        size_t dh = model->meta.dh;
        size_t nh = model->meta.nh;
        size_t hs = model->meta.hs;
        size_t nkvh = model->meta.nkvh;
        size_t dkvh = dh * nkvh;

        float eps = model->meta.epsilon;
        float theta = model->meta.theta;
        float scale = 1.0f / std::sqrt(static_cast<float>(dh));

        size_t seqlen = ntoken;
        size_t curlen = past_len + seqlen;

        auto k_cache = model->k_cache;
        auto v_cache = model->v_cache;

        std::vector<size_t> create_shape = {seqlen};
        //pos_ids for rope
        auto pos_ids = tensorCreate(create_shape.data(), 1, LLAISYS_DTYPE_I64, device_type, device_id);
        //fill pos_ids
        auto pos_ids_data = reinterpret_cast<int64_t *>(tensorGetData(pos_ids));
        for(size_t i = 0; i < seqlen; i++){
            pos_ids_data[i] = static_cast<int64_t>(i + past_len);
        }

        //load token_ids
        auto input_token_ids = tensorCreate(create_shape.data(), 1, LLAISYS_DTYPE_I64, device_type, device_id);
        tensorLoad(input_token_ids, token_ids);

        //input embed
        create_shape = {seqlen, hs};
        auto input_embed = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
        llaisysEmbedding(input_embed, input_token_ids, model->weights->in_embed);


        //transformer
        for(size_t i = 0; i < nlayer; i++){
            LOG(std::string("layer ") + std::to_string(i) + "/" + std::to_string(nlayer -1));
            /*
                normalization
            */
            auto attn_norm = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysRmsNorm(attn_norm, input_embed, model->weights->attn_norm_w[i], eps);
            // LOG("attn_norm");
            // tensorDebug(attn_norm);

            /*
                attention 
            */
            //q_proj
            auto q_proj = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysLinear(q_proj, attn_norm, model->weights->attn_q_w[i], model->weights->attn_q_b[i]);
            // LOG("q_proj");
            // tensorDebug(q_proj);

            //q_rope
            create_shape = {seqlen, nh, dh};
            q_proj = tensorView(q_proj, create_shape.data(), 3);
            auto q_rope = tensorCreate(create_shape.data(), 3, model->meta.dtype, device_type, device_id);
            llaisysROPE(q_rope, q_proj, pos_ids, theta);
            // LOG("q_rope");
            // tensorDebug(q_rope);

            //k_proj
            create_shape = {seqlen, dkvh};
            auto k_proj = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysLinear(k_proj, attn_norm, model->weights->attn_k_w[i], model->weights->attn_k_b[i]);
            // LOG("k_proj");
            // tensorDebug(k_proj);

            //k_rope
            create_shape = {seqlen, nkvh, dh};
             k_proj = tensorView(k_proj, create_shape.data(), 3);
            auto k_rope = tensorSlice(k_cache[i], 0, past_len, curlen);
            llaisysROPE(k_rope, k_proj, pos_ids, theta);
            // LOG("k_rope");
            // tensorDebug(k_rope);

            //v_proj in_place
            create_shape = {seqlen, dkvh};
            auto v_proj = tensorView(tensorSlice(v_cache[i], 0, past_len, curlen), create_shape.data(), 2);
            llaisysLinear(v_proj, attn_norm, model->weights->attn_v_w[i], model->weights->attn_v_b[i]);
            // LOG("v_proj");
            // tensorDebug(v_proj);

            //self attention
            create_shape = {seqlen, nh, dh};
            auto attn_val = tensorCreate(create_shape.data(), 3, model->meta.dtype, device_type, device_id);

            auto attn_k = tensorSlice(k_cache[i], 0, 0, curlen);
            auto attn_v = tensorSlice(v_cache[i], 0, 0, curlen);

            llaisysSelfAttention(attn_val, q_rope, attn_k, attn_v, scale);
            // LOG("self_attention");
            // tensorDebug(attn_out);


//o_proj
             create_shape = {seqlen, hs};
            auto o_proj = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            attn_val = tensorView(attn_val, create_shape.data(), 2);
            // LOG("attn_val");
            // tensorDebug(attn_val);
            llaisysLinear(o_proj, attn_val, model->weights->attn_o_w[i], nullptr);
            // LOG("o_proj");
            // tensorDebug(o_proj);

            //residual connection
            create_shape = {seqlen, hs};
            auto mlp_layer = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysAdd(mlp_layer, input_embed, o_proj);
            // LOG("mlp_layer");

            //mlp_norm
            create_shape = {seqlen, hs};
            auto mlp_norm = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysRmsNorm(mlp_norm, mlp_layer, model->weights->mlp_norm_w[i], eps);
            // LOG("mlp_norm");
            // tensorDebug(mlp_norm);

            /*
                mlp
            */

            //mlp_gate
            create_shape = {seqlen, di};
            auto mlp_gate = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysLinear(mlp_gate, mlp_norm, model->weights->mlp_gate_w[i], nullptr);
            // LOG("mlp_gate");
            // tensorDebug(mlp_gate);

             //mlp_up
            create_shape = {seqlen, di};
            auto mlp_up = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysLinear(mlp_up, mlp_norm, model->weights->mlp_up_w[i], nullptr);
            // LOG("mlp_up");
            // tensorDebug(mlp_up);

             //mlp_swiglu
            create_shape = {seqlen, di};
            auto mlp_swiglu = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysSwiGLU(mlp_swiglu, mlp_gate, mlp_up);
            // LOG("mlp_swiglu");
            // tensorDebug(mlp_swiglu);

//mlp_down
             create_shape = {seqlen, hs};
            auto mlp_down = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
            llaisysLinear(mlp_down, mlp_swiglu, model->weights->mlp_down_w[i], nullptr);
            // LOG("mlp_down");
            // tensorDebug(mlp_down);

            //residual connection
             llaisysAdd(input_embed, mlp_layer, mlp_down);
            // LOG("residual connection");
            // tensorDebug(mlp_layer);

            tensorDestroy(attn_norm);
            tensorDestroy(q_proj);
            tensorDestroy(k_proj);
            tensorDestroy(v_proj);
            tensorDestroy(attn_val);
            tensorDestroy(o_proj);
            tensorDestroy(mlp_layer);
            tensorDestroy(mlp_norm);
            tensorDestroy(mlp_gate);
            tensorDestroy(mlp_up);
            tensorDestroy(mlp_swiglu);
            tensorDestroy(mlp_down);
        }

        /*
            output
        */
        LOG("It is time to output");
        //rms_norm
        create_shape = {seqlen, hs};
        auto output = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
        llaisysRmsNorm(output, input_embed, model->weights->out_norm_w, eps);
        // tensorDebug(output);

        
        //output_embed
        create_shape = {seqlen, voc};
        auto output_embed = tensorCreate(create_shape.data(), 2, model->meta.dtype, device_type, device_id);
        llaisysLinear(output_embed, output, model->weights->out_embed, nullptr);
        // tensorDebug(output_embed);

//argmax
        auto output_ids = tensorSlice(output_embed, 0, seqlen - 1, seqlen);

        create_shape = {1};
        auto max_val = tensorCreate(create_shape.data(), 1, model->meta.dtype, device_type, device_id);
        auto max_ids = tensorCreate(create_shape.data(), 1, LLAISYS_DTYPE_I64, device_type, device_id);
        llaisysArgmax(max_ids, max_val, output_ids);

        int64_t token_id = *reinterpret_cast<int64_t *>(tensorGetData(max_ids));       

        //destroy
        tensorDestroy(pos_ids);
        tensorDestroy(input_token_ids);
        tensorDestroy(input_embed);
        tensorDestroy(output);
        tensorDestroy(output_embed);
        tensorDestroy(max_val);
        tensorDestroy(max_ids);

        return token_id;
    }
}

File: python/llaisys/models/qwen2.py
from typing import Sequence
from ..libllaisys import LIB_LLAISYS
from ..libllaisys import DeviceType, DataType, llaisysTensor_t
from .qwen2_config import *
from .utils import *

from pathlib import Path
import safetensors
import json
import logging
import sys

# 配置 logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    handlers=[
        logging.StreamHandler(sys.stdout),  # 明确指定使用重定向后的 stdout
        logging.FileHandler("qwen2.log", encoding="utf-8"),  # 文件也指定编码
    ],
)

logger = logging.getLogger(__name__)

load_qwen2(LIB_LLAISYS)


class Qwen2:
    def __init__(self, model_path, device: DeviceType = DeviceType.CPU):
        # TODO: Implement model constructor
        assert device == DeviceType.CPU, "Only CPU is supported for now"

        self.model_path = Path(model_path)
        self.device = device
        # self.model:LlaisysQwen2Model = {}

    def create_model(self, kv_cache_size: int = 0):
        # Load model meta
        with open(self.model_path / "config.json", "r") as f:
            self.config = json.load(f)
        # 将信息打印为绿色字体，并在两行横线之间打印

        self.meta = LlaisysQWen2Meta(
            dtype=torch_dtype_to_llaisys_dtype(self.config),
            nlayer=self.config["num_hidden_layers"],
            hs=self.config["hidden_size"],
            nh=self.config["num_attention_heads"],
            nkvh=self.config["num_key_value_heads"],
            dh=self.config["hidden_size"] // self.config["num_attention_heads"],
            di=self.config["intermediate_size"],
            maxseq=self.config["max_position_embeddings"],
            voc=self.config["vocab_size"],
            episilon=self.config["rms_norm_eps"],
            theta=self.config["rope_theta"],
            end_token=self.config["eos_token_id"],
        )

        # Load model weights
        state_dict = {}
        self.weights: LlaisysQWen2Weights = {}
        for file in sorted(self.model_path.glob("*.safetensors")):
            data_ = safetensors.safe_open(file, framework="torch", device="cpu")
            for name_ in data_.keys():
                state_dict[name_] = data_.get_tensor(name_)
                # TODO: load the weights to self.weights
        # 将信息打印为绿色字体，并在两行横线之间打印

        # c_int数组
        device_ids = (c_int * 1)(0)
        self.model = LIB_LLAISYS.llaisysQwen2ModelCreate(
            byref(self.meta),
            c_size_t(kv_cache_size),
            self.device,
            device_ids,
            len(device_ids),
        )

        self.weights = LIB_LLAISYS.llaisysQwen2ModelWeights(self.model)

        logger.info(
            f"Loading in_embed with shape: {state_dict['model.embed_tokens.weight'].shape}"
        )
        logger.info(f"in_embed dtype: {state_dict['model.embed_tokens.weight'].dtype}")
        LIB_LLAISYS.tensorLoad(
            getattr(self.weights.contents, "in_embed"),
            state_dict["model.embed_tokens.weight"].data_ptr(),
        )

        logger.info(
            f"Loading out_embed with shape: {state_dict['lm_head.weight'].shape}"
        )
        logger.info(f"out_embed dtype: {state_dict['lm_head.weight'].dtype}")
        LIB_LLAISYS.tensorLoad(
            getattr(self.weights.contents, "out_embed"),
            state_dict["lm_head.weight"].data_ptr(),
        )

        logger.info(
            f"Loading out_norm_w with shape: {state_dict['model.norm.weight'].shape}"
        )
        logger.info(f"out_norm_w dtype: {state_dict['model.norm.weight'].dtype}")
        LIB_LLAISYS.tensorLoad(
            getattr(self.weights.contents, "out_norm_w"),
            state_dict["model.norm.weight"].data_ptr(),
        )

        # refernce:https://github.com/liulog/llaisys/blob/main/python/llaisys/models/qwen2.py
        def load_layer_array(field_name, base_name):
            arr_ptr = getattr(self.weights.contents, field_name)
            arr_type = llaisysTensor_t * self.meta.nlayer
            arr = cast(arr_ptr, POINTER(arr_type)).contents

            for i in range(self.meta.nlayer):
                tensor_name = f"model.layers.{i}.{base_name}"
                LIB_LLAISYS.tensorLoad(arr[i], state_dict[tensor_name].data_ptr())

        load_layer_array("attn_norm_w", "input_layernorm.weight")
        load_layer_array("attn_q_w", "self_attn.q_proj.weight")
        load_layer_array("attn_q_b", "self_attn.q_proj.bias")
        load_layer_array("attn_k_w", "self_attn.k_proj.weight")
        load_layer_array("attn_k_b", "self_attn.k_proj.bias")
        load_layer_array("attn_v_w", "self_attn.v_proj.weight")
        load_layer_array("attn_v_b", "self_attn.v_proj.bias")
        load_layer_array("attn_o_w", "self_attn.o_proj.weight")

        load_layer_array("mlp_norm_w", "post_attention_layernorm.weight")
        load_layer_array("mlp_gate_w", "mlp.gate_proj.weight")
        load_layer_array("mlp_up_w", "mlp.up_proj.weight")
        load_layer_array("mlp_down_w", "mlp.down_proj.weight")

        # weights_to_print = LIB_LLAISYS.llaisysQwen2ModelWeights(self.model)
        # logger.info(weights_to_print)

    def generate(
        self,
        inputs: Sequence[int],
        max_new_tokens: int = None,
        top_k: int = 1,
        top_p: float = 0.8,
        temperature: float = 0.8,
    ):
        # TODO: Implement generate function
        logger.info(f"max_new_tokens: {max_new_tokens}")
        tokens = list(inputs)
        kv_cache_size = len(tokens) + max_new_tokens
        self.create_model(kv_cache_size)

        # prefill
        logger.info("prefill start")
        ntoken = len(tokens)
        token_ids = (c_int64 * ntoken)(*tokens)
        past_len = c_size_t(0)
        try:
            next_token = LIB_LLAISYS.llaisysQwen2ModelInfer(
                self.model, token_ids, ntoken, past_len
            )
            logger.info(f"prefill completed, next_token: {next_token}")
        except Exception as e:
            logger.error(f"Error during prefill: {e}")
            raise
        tokens.append(next_token)
        logger.info("prefill end")
        logger.info("decode start")
        # decode
        for _ in range(max_new_tokens - 1):
            logger.info(f"next_token: {next_token}")
            if next_token == self.meta.end_token:
                break
            logger.info(f"tokens: {tokens}")
            ntoken = 1
            token_ids = (c_int64 * 1)(next_token)
            past_len = c_size_t(len(tokens) - 1)
            logger.info(f"past_len: {past_len}")
            next_token = LIB_LLAISYS.llaisysQwen2ModelInfer(
                self.model, token_ids, ntoken, past_len
            )
            tokens.append(next_token)

        LIB_LLAISYS.llaisysQwen2ModelDestroy(self.model)

        return tokens
